Content.Server/Chemistry/EntitySystems/SolutionTransferSystem.cs

```diff
@@ -66,5 +67,91 @@ private void AddSetTransferVerbs(EntityUid uid, SolutionTransferComponent compon
                 args.Verbs.Add(verb);
             }
         }
+
+        /// <summary>
+        /// Transfer from a solution to another.
+        /// </summary>
+        /// <returns>The actual amount transferred.</returns>
+        public FixedPoint2 Transfer(EntityUid user,
+            EntityUid sourceEntity,
+            Solution source,
+            EntityUid targetEntity,
+            Solution target,
+            FixedPoint2 amount)
+        {
+            var transferAttempt = new SolutionTransferAttemptEvent(sourceEntity, targetEntity);
+
+            // Check if the source is cancelling the transfer
+            RaiseLocalEvent(sourceEntity, transferAttempt);
+            if (transferAttempt.Cancelled)
+            {
+                sourceEntity.PopupMessage(user, transferAttempt.CancelReason!);
+                return FixedPoint2.Zero;
+            }
+
+            if (source.DrainAvailable == 0)
+            {
+                sourceEntity.PopupMessage(user,
+                    Loc.GetString("comp-solution-transfer-is-empty", ("target", sourceEntity)));
+                return FixedPoint2.Zero;
+            }
+
+            // Check if the target is cancelling the transfer
+            RaiseLocalEvent(targetEntity, transferAttempt);
+            if (transferAttempt.Cancelled)
+            {
+                sourceEntity.PopupMessage(user, transferAttempt.CancelReason!);
+                return FixedPoint2.Zero;
+            }
+
+            if (target.AvailableVolume == 0)
+            {
+                targetEntity.PopupMessage(user,
+                    Loc.GetString("comp-solution-transfer-is-full", ("target", targetEntity)));
+                return FixedPoint2.Zero;
+            }
+
+            var actualAmount = FixedPoint2.Min(amount, FixedPoint2.Min(source.DrainAvailable, target.AvailableVolume));
+
+            var solutionSystem = Get<SolutionContainerSystem>();
+            var solution = solutionSystem.Drain(sourceEntity, source, actualAmount);
+            solutionSystem.Refill(targetEntity, target, solution);
+
+            return actualAmount;
+        }
+    }
+
+    /// <summary>
+    /// Raised when attempting to transfer from one solution to another.
+    /// </summary>
+    public sealed class SolutionTransferAttemptEvent : EntityEventArgs
+    {
+        public SolutionTransferAttemptEvent(EntityUid from, EntityUid to)
+        {
+            From = from;
+            To = to;
+        }
+
+        public EntityUid From { get; }
+        public EntityUid To { get; }
+
+        /// <summary>
+        /// Whether this transfer has been cancelled.
+        /// </summary>
+        public bool Cancelled { get; private set; }
```

Comment:
inherit `CancellableEntityEventArgs` instead of implementing that behavior yourself

Reply:
Can I still require an argument on `Cancel` if I do that? I think there should always be a reason.

Reply:
you can just set the field with the reason manually, no need to pass it into a function